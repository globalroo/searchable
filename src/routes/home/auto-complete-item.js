import React from "react";
import { MenuItem } from "@material-ui/core";
import { Movie, Tv, Person } from "@material-ui/icons/";
import { MEDIA_TYPE } from "src/tmdb-service/tmdb-api.js";
import { Typography } from "@material-ui/core";

const getAutoCompleteItemIcon = media_type => {
	switch (media_type) {
		case MEDIA_TYPE.MOVIE:
			return <Movie />;
		case MEDIA_TYPE.TV:
			return <Tv />;
		case MEDIA_TYPE.PERSON:
			return <Person />;
		default:
			return null;
	}
};

export const AutoCompleteItem = ({ item, index, highlighted, selected, ...props }) => (
	<MenuItem {...props} selected={highlighted} component="div">
		{getAutoCompleteItemIcon(item.media_type)}
		<Typography>{`${item.title || item.name}`}</Typography>
	</MenuItem>
);
