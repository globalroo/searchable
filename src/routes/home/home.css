/* CSS Grid styling via https://gridbyexample.com/patterns/ - Rachel Andrew @rachelandrew */

*,
*:before,
*:after {
    box-sizing: border-box;
}

h1,
p {
    margin: 0 0 1em 0;
}

.home-wrapper {
    max-width: 60%;
    margin: 0 20px;
    display: grid;
    grid-gap: 10px;
}

/* no grid support? */

.home-wrapper {
    display: flex;
    flex-wrap: wrap;
    padding-left: 10px;
    padding-right: 10px;
}

.home-wrapper {
    display: grid;
    margin: 1em auto;
    grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
    grid-auto-rows: minmax(150px, auto);
}

.home-panel {
    /* needed for the flex layout*/
    margin-left: 5px;
    margin-right: 5px;
    flex: 1 1 200px;
}

.home-header,
.home-footer {
    margin-top: 30px;
    margin-left: auto;
    margin-right: auto;
    width: 60%;
}

.home-wrapper>* {
    border-radius: 5px;
    padding: 10px;
    margin-bottom: 10px;
}

@supports (display: grid) {
    .home-wrapper>* {
        margin: 0;
    }
}

.image-tile {
    transition: transform .2s;
    filter: drop-shadow(0 0 0.05rem black);
}

.image-tile:hover {
    transform: scale(1.125);
    cursor: pointer;
}