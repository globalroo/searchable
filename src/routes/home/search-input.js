import React from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import Close from "@material-ui/icons/Close";
import MaterialDivider from "@material-ui/core/Divider";
import IconButton from "@material-ui/core/IconButton";
import InputBase from "@material-ui/core/InputBase";
import Paper from "@material-ui/core/Paper";
import SearchIcon from "@material-ui/icons/Search";

const styles = {
	root: {
		display: "flex",
		alignItems: "center"
	},
	input: {
		marginLeft: 10,
		flex: 1
	},
	iconButton: {
		padding: 10
	},
	divider: {
		width: 1,
		height: 28,
		margin: 4
	}
};

const ClearButton = ({ className, onClear }) => (
	<IconButton className={className} aria-label="Clear" onClick={onClear}>
		<Close />
	</IconButton>
);

const SearchButton = ({ className, onSearch }) => (
	<IconButton color="primary" className={className} aria-label="Search" onClick={onSearch}>
		<SearchIcon />
	</IconButton>
);

function BasicSearchInput({ classes, value, onChange, onClear, onSearch, ...props }) {
	return (
		<Paper className={classes.root}>
			<InputBase className={classes.input} value={value} onChange={onChange} {...props} />
			{value.length > 0 && <ClearButton onClear={onClear} className={classes.iconButton} />}
			<MaterialDivider className={classes.divider} />
			<SearchButton className={classes.iconButton} onSearch={onSearch} />
		</Paper>
	);
}

BasicSearchInput.propTypes = {
	classes: PropTypes.object.isRequired,
	onChange: PropTypes.func.isRequired,
	onSearch: PropTypes.func.isRequired,
	value: PropTypes.string.isRequired
};

export const SearchInput = withStyles(styles)(BasicSearchInput);
