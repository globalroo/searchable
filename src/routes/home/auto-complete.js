import React, { useEffect, useState } from "react";
import Downshift from "downshift";
import { Paper } from "@material-ui/core";
import { useDebounce } from "use-debounce";

import { getSearchMultiResults } from "src/tmdb-service/tmdb-api.js";
import { navigateTo } from "../../helpers/navigate";
import { AutoCompleteItem } from "./auto-complete-item";
import { SearchInput } from "./search-input";

const DEBOUNCE_TIME = 350; //ms
const MINIMUM_SEARCH = 3;

export const AutoComplete = ({ populate }) => {
	const [searchTerm, setSearchTerm] = useState("");
	const [isOpen, setIsOpen] = useState(false);
	const [results, setResults] = useState([]);

	const [debouncedSearchValue] = useDebounce(searchTerm, DEBOUNCE_TIME);

	useEffect(() => {
		setIsOpen(false);
		if (debouncedSearchValue.length === 0) {
			setResults([]);
			populate([]);
		} else if (debouncedSearchValue.length >= MINIMUM_SEARCH) {
			getSearchMultiResults(debouncedSearchValue).then(({ payload }) => {
				setIsOpen(true);
				setResults(() => payload.results.slice(0, 10));
			});
		}
	}, [debouncedSearchValue]);

	const searchValue = ({ target }) => setSearchTerm(target.value);

	const doSearch = () => {
		setIsOpen(false);
		populate(results);
	};

	const handleEnter = e => {
		if (e.key === "Enter") {
			doSearch();
		}
	};

	return (
		<Downshift
			inputValue={searchTerm}
			itemToString={item => (item ? item.title : "")}
			onChange={navigateTo}
			isOpen={isOpen}
		>
			{({ getInputProps, getItemProps, highlightedIndex, inputValue }) => (
				<div>
					<SearchInput
						label="Search for Movie, TV show or Actor"
						{...getInputProps({
							onChange: searchValue,
							onSearch: doSearch,
							onKeyDown: handleEnter,
							value: inputValue,
							placeholder: "Search Movies, TV Shows and Actors"
						})}
					/>
					{isOpen && (
						<Paper
							className="paper"
							square
							style={{ marginLeft: 0, marginRight: 5, marginTop: 2 }}
						>
							{results.map((item, index) => (
								<AutoCompleteItem
									highlighted={highlightedIndex === index}
									index={index}
									item={item}
									{...getItemProps({
										index,
										item,
										key: item.id
									})}
								/>
							))}
						</Paper>
					)}
				</div>
			)}
		</Downshift>
	);
};
