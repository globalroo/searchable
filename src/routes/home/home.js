import React, { useState } from "react";

import { AutoComplete } from "src/components/auto-complete";
import { Footer } from "src/components/footer";
import { TrendingSearches } from "src/components/trending";
import { AssetTile } from "src/components/asset-tile";

import "./home.css";

export const Home = () => {
	const [searchResults, setSearchResults] = useState([]);
	const populate = data => setSearchResults(data);

	return (
		<>
			<header className="home-header">
				<AutoComplete populate={populate} />
			</header>
			<div className="home-wrapper">
				{searchResults.map(tile => (
					<AssetTile key={tile.id} asset={tile} />
				))}
			</div>
			<footer className="home-footer">
				<TrendingSearches />
				<Footer />
			</footer>
		</>
	);
};
