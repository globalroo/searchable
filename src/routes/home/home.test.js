import React from "react";
import { render, fireEvent, waitForElement } from "react-testing-library";

import { Home } from "./home";

jest.mock("src/components/auto-complete", () => {
	return {
		AutoComplete: ({ populate }) => (
			<button
				onClick={() => {
					populate([{ id: 0 }, { id: 1 }]);
				}}
				data-testid="clicker"
			/>
		)
	};
});

jest.mock("src/components/footer", () => ({
	Footer: () => <div>Footer</div>
}));
jest.mock("src/components/trending", () => ({
	TrendingSearches: () => <div>Trending Searches</div>
}));
jest.mock("src/components/asset-tile", () => ({
	AssetTile: ({ asset }) => <div data-testid="test">{JSON.stringify(asset)}</div>
}));

describe("Home Component tests", () => {
	it("Renders a home with data component", async () => {
		const { container, getByTestId } = render(<Home />);
		const button = getByTestId("clicker");
		await fireEvent.click(button);
		await waitForElement(() => getByTestId("test"));
		expect(container).toMatchSnapshot();
	});
});
