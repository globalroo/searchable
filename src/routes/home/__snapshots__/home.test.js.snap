// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component tests Renders a home with data component 1`] = `
<div>
  <header
    class="home-header"
  >
    <button
      data-testid="clicker"
    />
  </header>
  <div
    class="home-wrapper"
  >
    <div
      data-testid="test"
    >
      {"id":0}
    </div>
    <div
      data-testid="test"
    >
      {"id":1}
    </div>
  </div>
  <footer
    class="home-footer"
  >
    <div>
      Trending Searches
    </div>
    <div>
      Footer
    </div>
  </footer>
</div>
`;
